### ISSUE TITLE
GAP records fail in various ways with long names

### ISSUE THREAD

**Issue by ChrisJefferson** [gap-system/gap#27]
Merging various bugs into one report. If you try to name members with long names (longer than about 1024 characters) into records, all kinds of things break, including:

- Enter break loop when you try to print them out
- Strings get truncated in RecNames
- If you put enough values in, older values will get over-written

To see this, type:

```
gap> r := rec();;
gap> r.(List([1..2000], x -> 'a')) := 1;;
gap> Length(RecNames(r)[1]); # should be 2000
1023
gap> r;
```
... hits an error because r doesn't have a member with name RecNames(r)[1]. I hit this because I was using records like dictionaries, as dictionaries were behaving badly.

I can see two routes of attack:

1) Enforce a limit on name of a record (probably 1023, same as identifiers), and error when anyone tries to insert such a longer name into a record.
2) Remove the hard-wired limitations on lengths of record members.

(1) is obviously easier, and I assume no-one else has ever hit this issue before. Happy to look at 2, obviously there is the risk of a slight efficiency loss

**Comment by ChrisJefferson** [comment]
I've been investigating what would be required for this:

- There is a low-level function `RNamName(char*)` which hashes the name of a string. This would have to be separated into two functions, `RNamName(char*)` which handles strings which are not in a GAP object (by far the most common case, and the only one used in packages), and `RNamNameObj(Obj)`, which handles strings in objects -- at the moment it does some internal copying to combine these cases

- There is a separate, working, limit in the GAP parser that identifiers cannot be longer than 1023 characters, so writing `r.aaa...<200 characters>...aaa := 1` is not going to work without massive changes. This is independent of `r.(x)` working for large `x` of course.

**Comment by markuspf** [comment]
I think we should refrain from using records as dictionaries/maps. I sounded like a slightly weird but workable idea at first. 

We can make the dictionary implementation to be in the datastructures package behave like a record by installing the correct operations, if you really want that behaviour (and I am not sure whether that is a good idea).

**Comment by ChrisJefferson** [comment]
I now plan on limiting the length, because I think allowing longer ones would interact poorly with the parser (you could make a record, print it out, then be unable to read it back in).

**Comment by markuspf** [comment]
It is documented in the reference manual that identifiers can be arbitrarily long, but only the first 1023 characters are significant.
This can be considered a bug in that of course your example should still work, I think?

**Comment by ChrisJefferson** [comment]
Interesting, because that's not how GAP works, so we need a fix in some direction :)
```
gap> xxx...<2000 characters>...xxx := 3;
Syntax error: Identifiers in GAP must consist of less than 1023 characters.
gap> r := rec();;
gap> r.xxx....<2000 characters>...x : = 3;
Syntax error: Identifiers in GAP must consist of less than 1023 characters.

**Comment by frankluebeck** [comment]
On Wed, Mar 11, 2015 at 04:20:26AM -0700, Christopher Jefferson wrote:
> Interesting, because that's not how GAP works, so we need a fix in some direction :)
> ```
> gap> xxx...<2000 characters>...xxx := 3;
> Syntax error: Identifiers in GAP must consist of less than 1023 characters.
> gap> r := rec();;
> gap> r.xxx....<2000 characters>...x : = 3;
> Syntax error: Identifiers in GAP must consist of less than 1023 characters.

This limitation is well known (or, at least I know about it since long). I
remember that I have introduced this error message, but at that time decided
that the limitation is not of enough practical relevance to put work into
this. And I cannot remember any report of a user stumbling over this.

Of course, it wouldn't hurt to clean this up. Since this limitation never lead 
to any practical problem I suggest to strictly limit identifiers to 1023 characters
(i.e. bytes) and to generate error messages when users try to use longer
ones (as the parser already does in the examples above).

**Comment by markuspf** [comment]
I was aware of the limit of 1023 bytes in an identifier, but only in the sense as documented in the reference manual. I never hit a case where I wanted to make an identifier of that length.

I agree that we should just limit lengths of identifiers to 1023 and update the documentation to reflect that.

**Comment by fingolfin** [comment]
I think limiting the identifier length like that is good idea, certainly simplifies life for us, and shouldn't hurt anybody.

As for interface for dictionaties: I think using the syntax proposed in PR #28 would be more flexible than (ab)using the record interface for that. Perhaps we can talk about this more next week?

**Comment by alex-konovalov** [comment]
This issue is closed with PR #59 merged.
